docs: add testing styleguide and amend readme and contributing guides (#2957)
Sage/carbon: docs/testing-styleguide.MD

#### Continuous Integration (CI)
The Travis CI service runs unit tests at every commit/push/pull request creation. You can manually run the tests with:
1.	`npm run lint` - runs lint test.
2.	`npm test` - runs jest tests.
3.	Results can be investigated from the terminal or from the Travis dashboard linked from the branch/pull request checks

### Functional Browser Testing
#### Manual Testing
Every new or amended component should be fully tested in either Storybook and/or [codesandbox.io](https://codesandbox.io) using functional testing, UI testing and internationalization testing techniques. We use [Jira](https://www.atlassian.com/software/jira) for issue and project tracking. 

##### Browser Support
Carbon is supported in the following browsers:
* Chrome
* Firefox
* Safari
* Edge Chromium

Components should be tested in all supported browsers unless the piece of functionality is not specifically related to a UI feature, e.g. decimal precision.

##### Jira
For every testable task, a test session needs to be created for each of the supported browsers. Results should be added to each of the conditions in the created test sessions with a pass `(/)` or fail `(x)` mark.

> Any issues found during testing should be raised as a "Defect" in Jira unless the ticket is already of a “Bug” type, in which case comments and screenshots/recordings/GIFs can be added to the existing ticket.

Reviewer: I think we should mention that if the bug is present in master and not a defect of this ticket, it should be its own `bug` ticket and shouldn't block this ticket.

Author: At line 80 I state "If the issue is present on the master branch you should first check if the issue has already been logged in GitHub. If it is present, it should be upvoted with a “+1” comment...Collaborators internal to Sage should also log a Jira ticket for the issue as a "Bug" type with the GitHub issue linked." I'll amend the sentence to read "Collaborators internal to Sage should also log a **_separate_** Jira ticket for the issue, **_not blocking this one_**, as a "Bug" type...". to specify that.

> If the issue is present on the master branch you should first check if the issue has already been logged in GitHub. If it is present, it should be upvoted with a “+1” comment. If it has not been logged, details of how to do this are contained in the [CONTRIBUTING.MD](https://github.com/Sage/carbon/blob/master/CONTRIBUTING.md) guide. Collaborators internal to Sage should also log a Jira ticket for the issue as a "Bug" type with the GitHub issue linked. However, there should be no link from the GitHub issue to the Jira ticket.

Reviewer: ```suggestion
If the issue is present on the master branch you should first check if the issue has already been logged in GitHub. If it is present, it should be upvoted with a “+1” comment. If it has not been logged, details of how to do this are contained in the [CONTRIBUTING.MD](https://github.com/Sage/carbon/blob/master/CONTRIBUTING.md) guide. Collaborators internal to Sage should also log a Jira ticket for the issue as a "Bug" type with the GitHub issue linked. The Jira ticket number should be included in the GitHub issue e.g. `FE-123`, but do not include the full JIRA URL.
```

##### Storybook

> Storybook forms our Carbon component library documentation. The component stories are written in `MDX` format containing as many examples as required to showcase our components and with minimal knobs for changing prop states. Every component should have at least one story (named `default`/`multiple`/`validation` etc.) located in the `Design System` directory with `Docs`, `Notes` and `Accessibility` sections. Where required, it should be possible to test all events, e.g. `onClick`, `drag` etc, in these stories. Results of events tests can be found in the `Actions` tab.

Reviewer 2: If I remember correctly, `knobs` should be **only** in `Test` folder to allow testing prop states, am I right ?

Author: Some knobs will need to be in `Design System` too for some components in order to showcase them, as discussed during the meeting with the team about the Applitools strategy. Three such examples exist currently, namely the `Drawer`, `Inline Inputs` and `Pager` components.

Author: Added a further comment stating components will be presented in set states.

> Storybook forms our Carbon component library documentation. The component stories are written in `MDX` format containing as many examples as required to showcase our components and with minimal knobs for changing prop states. Every component should have at least one story (named `default`/`multiple`/`validation` etc.) located in the `Design System` directory with `Docs`, `Notes` and `Accessibility` sections. Where required, it should be possible to test all events, e.g. `onClick`, `drag` etc, in these stories. Results of events tests can be found in the `Actions` tab.

Reviewer: ```suggestion
Storybook forms our Carbon component library documentation. The component stories are written in `MDX` format containing as many examples as required to showcase our components and with minimal knobs for changing prop states. Every component should have at least one story (named `default`/`multiple`/`validation` etc.) located in the `Design System` directory. Where required, it should be possible to test all events, e.g. `onClick`, `drag` etc, in these stories. Results of events tests can be found in the `Actions` tab.
```
MDX populates the `Docs` automatically. `Docs` is replacing `Notes` and `Accessibility` is automatically available in canvas view.

Where the stories in the main `Design System` documentation directory do not cover all scenarios required to fully test the component, additional stories must be created in the `Test` directory. For example, scenarios testing special characters, all combinations of size/type/state etc should have stories in the `Test` directory under the relevant component and named `visual`. A developer may need to assist with writing the additional required stories here.

To clarify the purpose of each directory,

| Directory     | Usage                                                                    |
|---------------|--------------------------------------------------------------------------|

> | Design System | User facing documentation, primary visual regression tests               |

Reviewer 2: I think that whole `visual regression` should be done in one place for the relevant component `Test > [component name] > visual`. We shouldn't split it into two separate places. I mean `Test > [component name] > visual` should contain as many examples as required for visual regression without any additional description

Author: I thought we agreed though that we didn't want to duplicate stories just for testing purposes? If we can do what visual testing we need in in the canvas in `Design System` then do we really need to duplicate the story just to run visual comparison testing? May be I missed something during the team call about this so happy to change it if that is the case.

Author: During the team meeting we had about the Applitools strategy, we discussed how that approach would mean duplicating stories just for visual testing purposes. We agreed to do visual testing against the Docs in stories in `Design System` unless we required specific test scenarios that were not covered in `Design System` or if we find that the length of the page in Docs is too long and contains too much text and white space. Happy to discuss it again though? May be when we look at the numbers of components requiring visual testing to get an estimate of required snapshots for the next meeting with Applitools we can get an idea of how much duplication there'd be if we took this approach? If we change our processes then I can change this doc then.

Author: Since decided to place all visual testing stories in the `Test` directory as you mentioned above. Updated this doc to reflect those discussions.

> | Design System | User facing documentation, primary visual regression tests               |

Reviewer: We need to state that Cypress regression (non-visual) should be exercising the Design System stories where possible. `Test` should just be if we need something that the documentation doesn't show (this shouldn't be often).

Author: I'll amend this line (91) in the table to read "User facing documentation, Cypress regression and visual tests". Line 85 just above explains that "Where the stories in the main `Design System` documentation directory do not cover all scenarios required to fully test the component, additional stories must be created in the `Test` directory."

Depending on how Cypress was run, you will find test results in either the Cypress UI, the console or in GitHub Actions linked from the build checks in GitHub. For further information see the [Getting Started Guide](https://github.com/Sage/carbon/blob/master/cypress/README.md).

##### Applitools
###### Integration with Cypress
Instructions for adding Applitools to an existing Cypress project can be found [here](https://applitools.com/tutorials/cypress.html). Applitools is integrated with Cypress by means of various configuration and settings in the `cypress/support/index.js` file, `applitools.config.js` file, and with additional feature files and corresponding step definitions of those components/stories where we want to do visual testing. When the relevant tests are run from Cypress, snapshots will be taken where specified in the feature files. We currently run visual tests in Chrome browser only.

An example of a feature file for a visual testing step follows. The `Then` step is where Applitools will take a snapshot:

```
Feature: Button all buttons component
  I want to check that all examples of Button component render correctly per theme

  @positive
  @applitools

>   Scenario Outline: Check that all buttons render correctly with theme set to <theme>

Reviewer 3: there is no such test in repo :(

Author: I know, it was just meant to be an example of how one could be written. I'll change it to an actual example to avoid confusion.

Author: I used the `Button component - primary type` spec as an example here.

      }, wrapper, { modifier: ':focus' });
    });
});
```

#### Snapshots
Snapshots render a string output of what the DOM receives. The first time the test is run, a snapshot file is generated and stored in the `__snapshots__` directory within each component in the `/src/components` directory. The next time a test is run, if nothing has changed, the test passes. However, if we change something, the test fails and a diff is indicated in the output. You can update snapshots easily using the `jest -u` command although you must take due care and consideration to ensure the changes are correct before simply updating them.

You should use snapshots in tests if amendments are being made to non-styled components to check styling is applied, but where existing and new styled components are concerned the use of snapshots is left up to the developer to decide if there is value. If you do want to use them, ensure they are small, focussed and effective.

#### Continuous Integration (CI)
The Travis CI service runs unit tests at every commit/push/pull request creation. You can manually run the tests with:
1.	`npm run lint` - runs lint test.
2.	`npm test` - runs jest tests.

> 3.	Results can be investigated from the terminal or from the Travis dashboard linked from the branch/pull request checks

Reviewer 3: not applied yet :(

> 3.	Results can be investigated from the terminal or from the Travis dashboard linked from the branch/pull request checks

Author: Done.

If the issue is present on the master branch you should first check if the issue has already been logged in GitHub. If it is present, it should be upvoted with a “+1” comment. If it has not been logged, details of how to do this are contained in the [CONTRIBUTING.MD](https://github.com/Sage/carbon/blob/master/CONTRIBUTING.md) guide. Collaborators internal to Sage should also log a separate Jira ticket for the issue, not blocking this one, as a "Bug" type with the GitHub issue linked. The Jira ticket number should be included in the GitHub issue, e.g. `FE-1234`, but do not include the full Jira URL.

##### Storybook
Storybook forms our Carbon component library documentation. The component stories are written in `MDX` format containing as many examples as required in set states to showcase our components. Every component should have at least one story (named `default`/`multiple`/`validation` etc.) located in the `Design System` directory. Where required, it should be possible to test all events, e.g. `onClick`, `drag` etc., in these stories. Results of events tests can be found in the `Actions` tab.

> Where the stories in the main `Design System` documentation directory do not cover all scenarios required to fully test the component, additional stories (named `basic`/`grouped`/`styles overridden` etc.) must be created in the `Test` directory. For example, scenarios testing special characters, all combinations of size/type/state etc. should have stories in the `Test` directory under the relevant component. Stories for visual testing should also be located here and named `visual`. These stories should be written with the components in set states where possible to avoid the use of knobs although they may be required in some cases, like to allow different text strings to be entered for instance. A developer may need to assist with writing the additional required stories here.

Reviewer 3: ```suggestion
Where the stories in the main `Design System` documentation directory do not cover all scenarios required to fully test the component, additional stories (named `basic`/`grouped`/`styles overridden` etc.) must be created in the `Test` directory. For example, scenarios testing special characters, all combinations of size/type/state etc. should have stories in the `Test` directory under the relevant component. Stories for visual testing should also be located here in `Test` directory and named `visual`. These stories should be written with the components in set states where possible to avoid the use of knobs although they may be required in some cases, like to allow different text strings to be entered for instance. A developer may need to assist with writing the additional required stories here.
```

Where functionality is already tested in unit testing, this does not need to be repeated in assertions in Cypress tests, with the exception of events tests which should be repeated here in order to test actions in a manner more closely resembling that of a user in a browser.

The visual comparison testing tool [Applitools](https://applitools.com) is used to test for UI regressions at each build by comparing new snapshots of the component preview iFrame with previous baseline ones. Applitools Eyes commands are included in Cypress tests where required. Visual test files should be stored in the `/visual` directory.

##### Cypress File Structure
```
.
├── cypress
│ ├── fixture
│ ├── features
│ │   ├── regression
│ │       ├── accessibility
│ │       ├── designSystem
│ │       ├── experimental
│ │       ├── test
│ │       ├── themes

> │ │       ├── validations

Reviewer 3: Please delete this line, because all `validation` tests were moved to the `visual` and deleted from cypress `regression` suite.

`validations`

Author: I took this straight from the original Cypress README.md so assumed everything on there was current information.

Author: I'll update this document once the changes are in `master`.

  * Step definitions are created per component as `component-steps.js` files and stored in `carbon/cypress/support/step-definitions`. They should follow the `component.feature` files exactly;
  * Common steps (`common-steps.js`) are used for all common actions.

Some common test steps:
  * I set `<knobs field>` to `<parameter>`;
  * I select `<knobs field>` to `<parameter>`;
  * I check/uncheck `<knobs field>` checkbox;
  * I open `<name>` component page;
    * To open different stories on Storybook;
      * I open `<name>` component page `default`/`basic`/`with button`/`legacy spinner`/`iframe`/`with button page in iframe`/`multiple`/`validations`.

##### Scenario tags
Test scenarios in feature files can be tagged to enable a subset of scenarios to be run, ignored or identified in some manner. Use the following tags:

1.	`@positive` for a happy path.
2.	`@negative` for the negative scenario.
3.	`@ignore` for temporarily ignored scenarios.
4.	`@[bug-number]` for example `@FE-1234` - use this tag after @ignore to explain why the test is ignored.
5.	`@ignore` and `@[bug-number]` should be removed after the bug fix.
6.	`@accessibility` tests verify accessibility violations.

> 7.	`@validations` tests verify validation components.

Reviewer 3: also `@validations` needs to be removed from here as I mentioned above.

Author: Again, I took this straight from the Cypress README.md and assumed it was current information.

Author: I'll update this document once the changes are in `master`.

Always use unique selectors to locate elements with the order below:
1.	`[data-component="unique-tag"]` for component and `[data-element="unique-tag"]` for component's element
2.	`#id`
3.	`.class-name`

###### How to name locators:
*  Component on preview - `componentPreview`
*  Knobs - `componentKnobs`

An example of an `index.js` file:
```
import { BUTTON_SUBTEXT_PREVIEW, BUTTON_DATA_COMPONENT_PREVIEW } from './locators';

// component preview locators
export const buttonSubtextPreview = () => cy.iFrame(BUTTON_SUBTEXT_PREVIEW);
export const buttonDataComponent = () => cy.iFrame(BUTTON_DATA_COMPONENT_PREVIEW);

// component preview locators in no IFrame
export const buttonDataComponentNoIFrame = () => cy.get(BUTTON_DATA_COMPONENT_PREVIEW);
```

An example of a corresponding `locators.js` file:

```
// component preview locators
export const BUTTON_SUBTEXT_PREVIEW = '[data-element="subtext"]';
export const BUTTON_DATA_COMPONENT_PREVIEW = '[data-component="button"]';
```

Other support files include the `helper.js` file which contains functions such as that used to prepare the URL of the component page and the control of draggable components.

##### Test Results

> Depending on how Cypress was run, you will find test results in either the Cypress UI, the console or in GitHub Actions linked from the build checks in GitHub. For further information see the [Getting Started Guide](https://github.com/Sage/carbon/blob/master/cypress/README.md).

Reviewer 2: ```suggestion
Depending on how Cypress was run, you will find test results in either the Cypress Test Runner UI, the console or in GitHub Actions linked from the build checks in GitHub. For further information see the [Getting Started Guide](https://github.com/Sage/carbon/blob/master/cypress/README.md).
```

##### Applitools
###### Integration with Cypress

> Instructions for adding Applitools to an existing Cypress project can be found [here](https://applitools.com/tutorials/cypress.html). Applitools is integrated with Cypress by means of various configuration and settings in the `cypress/support/index.js` file, `applitools.config.js` file, and with additional feature files and corresponding step definitions of those components/stories where we want to do visual testing. When the relevant tests are run from Cypress, snapshots will be taken where specified in the feature files. We currently run visual tests in Chrome browser only.

Reviewer 3: I think we need also to mention about using only `chrome` as a browser in `applitools` checks.
